### DISCUSSION TITLE
Pico thread dies after setting LED

### DISCUSSION THREAD

**Discussion by gwynge** [micropython/micropython#9595] [General]
I created a simple bit of code to detect a button and when it's pressed light the onboard LED.

As I wanted to explore threading I used a separate thread for this. This is the code I used to try and prove the concept. Whilst I'm using a button it's wired into an ADC (I was previously using a photoresistor and didn't bother changing the code, but same deal)

```python
led = Pin('LED',Pin.OUT)

ledOn = 0
adc = ADC(26)

def check_button():
    global ledOn
    print("check button")
    while True:
        print('flash check')
        val = adc.read_u16()
        if val < 40000:
            print('LED ON')
            led.value(1)
            ledOn = 1
        else:
            led.value(0)
            if ledOn:
                print('LED OFF')
                ledOn = 0
        print('sleep')
        time.sleep_ms(1000)
    print('no longer true')    

try:
    _thread.start_new_thread(check_button,())
    while True:
        print('main loop')
        time.sleep(2)
except KeyboardInterrupt as ex:
    print('Stop requested')

finally:
    print("Stopped")
```

Point is - when I run this you get the messages out showing it working. Until you press the button and it lights the LED. At that point the thread crashes. Comment the LED actions out and it works perfectly,

So why does accessing the LED kill the thread? Is this expected? Have I missed something in the documentation? Or is this a bug of some kind. _thread is marked as very experimental,

rp2-pico-w-20221011-unstable-v1.19.1-532-g8e912a501

**Comment by jimmo** [comment]
What version of MicroPython are you using? There have been numerous fixes to threading.

FWIW, your exact code snippet works for me on the latest firmware from https://micropython.org/download/rp2-pico/ e.g. [v1.19.1-543-gab317a0d6 (2022-10-11) .uf2](https://micropython.org/resources/firmware/rp2-pico-20221011-unstable-v1.19.1-543-gab317a0d6.uf2)

**Comment by gwynge** [comment]
I'm on a pico-w and I've just upgraded to rp2-pico-w-20221011-unstable-v1.19.1-543-gab317a0d6

And tried again.. this is the output I get 
```
main loopcheck button

flash check
sleep
flash check
sleep
main loop
flash check
sleep
flash check
sleep
main loop
flash check
sleep
flash check
LED ON
main loop
main loop
main loop
main loop
```

There is a difference between the boards inasmuch as the onboard LED is addressed differently..

**Reply by jimmo** [reply]
Ohhh... missed it was on a Pico W.

I have a suspicion that if you try turning on then off the LED at the start of the program, before you start the thread, then it will fix the issue.

On the Pico W, the LED is accessed through the WiFi chip (they ran out of pins, so had to use some of the extra GPIO on the WiFi chip). This gets initialised the first time you access one of these pins. So moving that to the main thread may help.

Let me know if that works and I'll raise a bug.

**Reply by gwynge** [reply]
I tried that and it didn't work, Well, if I set the LED on first and the thread first tried to set it on,,, that worked, but when it tried to turn it off again it failed. And vice versa, if I turned it on and off and then started the thread and if it tried to turn off it worked but turning on failed. So basically if it had to change value it failed.

[..edited because I wrote some rubbish before. led.on is not the same as led.on()!..]

**Reply by jimmo** [reply]
Was just able to test this -- I can't get it to work at all from the thread (with any of on/off/toggle/value). It appears the driver hangs waiting for a response from the wifi chip. I will raise an issue with some notes.

**Reply by jimmo** [reply]
https://github.com/micropython/micropython/issues/9597

**Reply by gwynge** [reply]
Thanks :o)
